quicklens/src/main/scala-3/com.softwaremill.quicklens/package.scala

```diff
@@ -0,0 +1,290 @@
+    import qctx.reflect.*
+
+    val focuses = focusesExpr match {
+      case Varargs(args) => args
+    }
+
+    val modF1 = fromPathModify(modifyImpl(obj, focus))
+    val modF = to[(A => A), S] { (mod: Expr[A => A]) =>
+      focuses.foldLeft(from[(A => A), S](modF1).apply(mod)) {
+        case (objAcc, focus) =>
+          val modCur = fromPathModify(modifyImpl(objAcc, focus))
+          from[(A => A), S](modCur).apply(mod)
+      }
+    }
+
+    toPathModify(obj, modF)
+  }
+
+  def modifyImpl[S, A](obj: Expr[S], focus: Expr[S => A])(using qctx: Quotes, tpeS: Type[S], tpeA: Type[A]): Expr[PathModify[S, A]] = {
+    import qctx.reflect.*
+
+    enum PathSymbol:
+      case Field(name: String)
+      case FunctionDelegate(name: String, givn: Term, typeTree: TypeTree, args: List[Term])
+
+    def toPath(tree: Tree): Seq[PathSymbol] = {
+      tree match {
+        case Select(deep, ident) =>
+          toPath(deep) :+ PathSymbol.Field(ident)
+        case Apply(Apply(Apply(TypeApply(Ident(s), typeTrees), idents), args), List(ident: Ident)) =>
+          idents.flatMap(toPath) :+ PathSymbol.FunctionDelegate(s, ident, typeTrees.last, args)
+        case a@Apply(Apply(TypeApply(Ident(s), typeTrees), idents), List(ident: Ident)) =>
+          idents.flatMap(toPath) :+ PathSymbol.FunctionDelegate(s, ident, typeTrees.last, List.empty)
+        case Apply(deep, idents) =>
+          toPath(deep) ++ idents.flatMap(toPath)
+        case i: Ident if i.name.startsWith("_") =>
+          Seq.empty
+        case _ =>
+          report.error(shapeInfo)
+          ???
+      }
+    }
+
+    def termMethodByNameUnsafe(term: Term, name: String): Symbol = {
+      term.tpe.typeSymbol.memberMethod(name).head
+    }
+
+    def termAccessorMethodByNameUnsafe(term: Term, name: String): (Symbol, Int) = {
+      val caseFields = term.tpe.typeSymbol.caseFields
+      val idx = caseFields.map(_.name).indexOf(name)
+      (caseFields.find(_.name == name).get, idx+1)
+    }
+
+    def mapToCopy(mod: Expr[A => A], objTerm: Term, path: Seq[PathSymbol]): Term = path match
+      case Nil =>
+        val apply = termMethodByNameUnsafe(mod.asTerm, "apply")
+        Apply(Select(mod.asTerm, apply), List(objTerm))
+      case (field: PathSymbol.Field) :: tail =>
+        val copy = termMethodByNameUnsafe(objTerm, "copy")
+        val (fieldMethod, idx) = termAccessorMethodByNameUnsafe(objTerm, field.name)
+        val namedArg = NamedArg(field.name, mapToCopy(mod, Select(objTerm, fieldMethod), tail))
+        val fieldsIdxs = 1.to(objTerm.tpe.typeSymbol.caseFields.length)
+        val args = fieldsIdxs.map { i =>
+          if(i == idx) namedArg
+          else Select(objTerm, termMethodByNameUnsafe(objTerm, "copy$default$" + i.toString))
+        }.toList
+        Apply(
+          Select(objTerm, copy),
+          args
+        )
+      case (f: PathSymbol.FunctionDelegate) :: tail =>
+        val defdefSymbol = Symbol.newMethod(
```

Comment:
could you maybe comment on why we need to generate an anonymous function here?
some pseudo-code in the comments which would show the shape of what we're generating would be very helpful

Reply:
Added some pseudo-code of generated code, see if it looks clearer now.

Reply:
Thanks! Yes, much clearer. Maybe I would also add some comment on what to `FunctionDelegate` itself - what it is, and what it represents in terms of quicklens.